benchexec/containerized_tool.py

```diff
@@ -57,13 +57,13 @@ def __init__(self, tool_module, config):
 
         container_options = containerexecutor.handle_basic_container_args(config)
         temp_dir = tempfile.mkdtemp(prefix="Benchexec_tool_info_container_")
-
+        self.container_id = None
         # Call function that loads tool module and returns its doc
         try:
-            self.__doc__ = self._pool.apply(
+            self.__doc__, self.container_id = self._pool.apply(
                 _init_container_and_load_tool,
-                [tool_module, temp_dir],
-                container_options,
+                [initializer] + self.mk_args(tool_module, temp_dir),
+                self.mk_kwargs(container_options),
```

Comment:
This
- adds code that is specific to the new subclass to the parent class (`container_id`),
- needs two hooks for letting the subclass overwrite the behavior, and
- still has code that is not needed for the new subclass (`temp_dir`, `container_options`).

It looks like this is not the best way to distribute the code over the inheritance hierarchy.

I would recommend that the whole `apply` call plus setup and cleanup is moved into a method that can be overwritten. `_init_container_and_load_tool` can be extended to pass through the result value, but it does not need to know anymore that it is the `container_id`.

The split into three classes looks similarly not-so-useful: We have the base class, which contains the common code but also specific code for the regular containerized tool-info module. Then we have `ContainerizedTool`, which just adds one little bit of information (what the `initializer` is). Either the base class should have nothing that is specific, or we just have two classes, the class that we had so far plus one new subclass. I would go with the latter choice.

Reply:
Thanks for the feedback. I was so focused on the arguments to the pool that I didn't realize, that i could just also call pool.apply from the subclass.

I reverted to the main ContainerizedTool again, this time just replacing the setup call to pool with a method `_setup_container`. The _init_and_load still resturns a tuple, but this is just ignored in the ContainerizedTool class. Only the subclass saves and uses the info.

Reply:
Why not move the setup and cleanup (`handle_basic_container_args`, `mkdtemp`, etc.) into `_setup_container` as well like I recommended?

Reply:
I misunderstood your recommendation.
There is a `finally:` that handles removal of the tempdir. Is it save to move this finally to the `_setup_container aswell`?

Reply:
If your code does not need the temp directory, you can move its creation into `_setup_container`. And then you can also move the cleanup. And to me it looks as if your code does not need it.